subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/internal/component/SelectionFailureHandlerIntegrationTest.groovy

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.internal.component
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.test.fixtures.dsl.GradleDsl
+
+import static org.gradle.test.fixtures.dsl.GradleDsl.*
+
+/**
+ * These tests demonstrate the behavior of the [SelectionFailureHandler] when a project has various
+ * variant selection failures.
+ */
+class SelectionFailureHandlerIntegrationTest extends AbstractIntegrationSpec {
+    def "demonstrate project ambiguous variant selection failure #dsl"() {
+        (dsl == GROOVY ? buildFile : buildKotlinFile) << """
+            ${setupAmbiguousVariantSelectionFailureForProject(dsl)}
+
+            ${forceConsumerResolution(dsl)}
+        """
+
+        expect:
+        fails "help", "outgoingVariants", "resolvableConfigurations"
```

Comment:
🤔 Addressing the comment above (about resolution during configuration) will probably address this, but we should be clear which task we're running here and which we think is going to fail.